File: 032. Array Of Products/ArrayOfProductsOnOnSimplified.java
import java.util.*;

class Program {

    // O(n) time | O(n) space - where n is the length of the input array
    public static int[] arrayOfProducts(int[] array) {
        int[] products = new int[array.length];

        int leftRunningProduct = 1;
        for (int i = 0; i < array.length; i++) {
            products[i] = leftRunningProduct;
            leftRunningProduct *= array[i];
        }

        int rightRunningProduct = 1;
        for (int i = array.length - 1; i >= 0; i--) {
            products[i] *= rightRunningProduct;
            rightRunningProduct *= array[i];
        }

        return products;
    }
}

File: 032. Array Of Products/ArrayOfProductsOnOnSimplified.kt
package com.algoexpert.program

// O(n) time | O(n) space - where n is the number of elements in the input array
fun arrayOfProducts(array: List<Int): List<Int> {
    val products = MutableList(array.size) {1}

    var leftRunningProduct = 1
    for (i in 0..array.size - 1) {
        products[i] = leftRunningProduct
        leftRunningProduct *= array[i]
    }

    var rightRunningProduct = 1
    for (i in array.size-1 downTo 0) {
        products[i] *= rightRunningProduct
        rightRunningProduct *= array[i]
    }

    return products
}
